### PR TITLE
Macro for expanding interpolator

### LINKED ISSUE #68: Macro for expanding the interpolator definition
For a "compile time" version of the interpolator, the user needs to define his own interpolator in a different compile unit.

*TBD*

### PR DESCRIPTION
**Pull request by fedefernandez**
Fixes #68 

It also changes the `MetadataSchemaProvider` implementation, removing the algebra dependency because we're going always to interpret to `Try`. In this way is better to just use the synchronized `cluster` API.

The metadata test will be addressed in #88

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -31,4 +31,14 @@ lazy val core = project
...
\ No newline at end of file
+  .settings(addCompilerPlugin("org.scalameta" % "paradise" % "3.0.0-M10" cross CrossVersion.full))
+
+lazy val `macros-tests` = project
+  .in(file("macros-tests"))
+  .settings(moduleName := "frees-cassandra-macros-tests")
+  .settings(scalaMetaSettings)
+  .settings(libraryDependencies ++= testDependencies)
+  .settings(libraryDependencies += "org.apache.cassandra" % "cassandra-all" % "3.11.0" % "test")
+  .settings(addCompilerPlugin("org.scalameta" % "paradise" % "3.0.0-M10" cross CrossVersion.full))
```
Comment (juanpedromoreno) [inline build.sbt]: I think this wouldn't be necessary if you are using `scalaMetaSettings`.

File: core/src/main/scala/query/interpolator/MacroInterpolator.scala [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  case object SchemaFile extends ValidatorType
+  case object Metadata extends ValidatorType
+
+  def validatorBlock(typeName: Type.Name, path: String, validator: ValidatorType): Term.Block = {
+
+    def schemaFileValidator(typeName: Type.Name, schemaPath: String) =
+      q"""
+          val tryMonadError: MonadError[Try, Throwable] = cats.instances.try_.catsStdInstancesForTry
+          val schemaProvider: SchemaDefinitionProvider[Try] =
+            TroySchemaProvider[Try](${Term.Name(typeName.value)}.getClass.getResourceAsStream(${Lit.String(schemaPath)}))(tryMonadError)
```
Comment (juanpedromoreno) [inline core/src/main/scala/query/interpolator/MacroInterpolator.scala]: What would happen if the resource is not in the classpath?

File: core/src/main/scala/query/interpolator/MacroInterpolator.scala [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+          TroySchemaValidator.instance(tryMonadError, schemaProvider)
+       """
+
+    def metadataValidator(typeName: Type.Name, schemaPath: String) =
+      q"""
+          val tryMonadError: MonadError[Try, Throwable] = cats.instances.try_.catsStdInstancesForTry
+          val schemaProvider: SchemaDefinitionProvider[Try] =
+            MetadataSchemaProvider.metadataSchemaProvider[Try](${Term.Name(typeName.value)}.getClass.getResourceAsStream(${Lit.String(schemaPath)}))(tryMonadError)
```
Comment (juanpedromoreno) [inline core/src/main/scala/query/interpolator/MacroInterpolator.scala]: Same here

File: core/src/main/scala/query/interpolator/MacroInterpolator.scala [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+     """
+
+  class SchemaFileInterpolator(schemaPath: String) extends scala.annotation.StaticAnnotation {
+
+    inline def apply(defn: Any): Any = meta {
+      val arg = this match {
+        case q"new $_(${Lit(argument: String)})" if argument.nonEmpty =>
+          argument
+        case _ =>
+          abort("You must provide a valid schema path")
```
Comment (juanpedromoreno) [inline core/src/main/scala/query/interpolator/MacroInterpolator.scala]: `@SchemaFileInterpolator annotation requires a valid schema path`?

File: macros-tests/src/test/scala/interpolator/MetadataInterpolatorTest.scala [hunk 5]
```diff
@@ -0,0 +1,74 @@
+/*
```
Comment (juanpedromoreno) [inline macros-tests/src/test/scala/interpolator/MetadataInterpolatorTest.scala:1]: Maybe we could move these tests to integration tests.

**Review by juanpedromoreno — APPROVED** [review]
Minor comments but it looks nice.

giphy 27

File: macros-tests/src/test/scala/interpolator/MetadataInterpolatorTest.scala (hunk 5 above)
Comment (fedefernandez) [reply macros-tests/src/test/scala/interpolator/MetadataInterpolatorTest.scala:1]: I'll do it as part of #88

**Review by juanpedromoreno — APPROVED** [review]
Latest changes look great, thanks!

**Pull request merged**